Support Enums with Generic Type Parameters

I love the idea for this crate! Unfortunately when I tried to use it I ran into some trouble because the enum I wanted to apply `DocumentedVariants` to has a generic type parameter `L`:

```rust
#[derive(.., Clone, .., documented::DocumentedVariants)]
pub enum Instr<L> {
    /// Dispatch on the type of the value pointed to by `X1`.
    SwitchOnTerm { .. }
    // ..
}
```

The error I get says:

```
error[E0107]: missing generics for enum `Instr`
  --> src\bc\instr.rs:93:10
   |
93 | pub enum Instr<L> {
   |          ^^^^^ expected 1 generic argument
   |
note: enum defined here, with 1 generic parameter: `L`
  --> src\bc\instr.rs:93:10
   |
93 | pub enum Instr<L> {
   |          ^^^^^ -
help: add missing generic argument
   |
93 | pub enum Instr<L><L> {
   |               +++

```

I figured out a workaround though.

1. In my case I could assign the type `()` as a default for `L`:
   
    ```rust
    pub enum Instr<L = ()> {
    ```
2. Then I wrote a helper impl:
    ```rust
    impl<L: Clone> Instr<L> {
        pub fn doc_comment(&self) -> Option<&'static str> {
            let mapped: Instr<()> = self.clone().map_lbl(|_| ()); // Throw away `L`, make it `()`.
            documented::DocumentedVariants::get_variant_docs(&mapped).ok()
        }
    }
    ```
    Unfortunately this requires that `L` be `Clone` (still ok for my use case, but unsatisfying).

    Here `map_lbl` maps the `L` parameter to a new type. It's signature looks like this:
    ```rust
    impl<L> Instr<L> {
        pub fn map_lbl<M>(self, f: impl Fn(L) -> M) -> Instr<M> { .. }
    }
    ```

I'm not sure how the derive macro works, but it seems like handling generic type parameters just hasn't been implemented yet. Would this be easy to fix?

--- a/documented-derive/src/lib.rs
+++ b/documented-derive/src/lib.rs
@@ -14,6 +14,9 @@ fn crate_module_path() -> Path {
 pub fn documented(input: TokenStream) -> TokenStream {
     let input = parse_macro_input!(input as DeriveInput);
 
+    let ident = &input.ident;
+    let (impl_generics, ty_generics, where_clause) = input.generics.split_for_impl();
+
     let doc_comments = match get_comment(&input.attrs) {
         Ok(Some(doc)) => doc,
         Ok(None) => {
@@ -24,11 +27,9 @@ pub fn documented(input: TokenStream) -> TokenStream {
         Err(e) => return e.into_compile_error().into(),
     };
 
-    let ident = input.ident;
-
     quote! {
         #[automatically_derived]
-        impl documented::Documented for #ident {
+        impl #impl_generics documented::Documented for #ident #ty_generics #where_clause {
             const DOCS: &'static str = #doc_comments;
         }
     }
@@ -39,6 +40,9 @@ pub fn documented(input: TokenStream) -> TokenStream {
 #[proc_macro_derive(DocumentedFields)]
 pub fn documented_fields(input: TokenStream) -> TokenStream {
     let input = parse_macro_input!(input as DeriveInput);
+
+    let ident = &input.ident;
+    let (impl_generics, ty_generics, where_clause) = input.generics.split_for_impl();
 
     let fields_doc_comments = {
         let fields_attrs: Vec<(Option<syn::Ident>, Vec<Attribute>)> = match input.data.clone() {
@@ -66,8 +70,6 @@ pub fn documented_fields(input: TokenStream) -> TokenStream {
         }
     };
 
-    let ident = input.ident;
-
     let (field_idents, field_comments): (Vec<_>, Vec<_>) = fields_doc_comments.into_iter().unzip();
 
     // quote macro needs some help with `Option`s
@@ -91,10 +93,10 @@ pub fn documented_fields(input: TokenStream) -> TokenStream {
 
     quote! {
         #[automatically_derived]
-        impl documented::DocumentedFields for #ident {
+        impl #impl_generics documented::DocumentedFields for #ident #ty_generics #where_clause {
             const FIELD_DOCS: &'static [Option<&'static str>] = &[#(#field_comments_tokenised),*];
 
-            fn __documented_get_index<T: AsRef<str>>(field_name: T) -> Option<usize> {
+            fn __documented_get_index<__Documented_T: AsRef<str>>(field_name: __Documented_T) -> Option<usize> {
                 use #documented_module_path::_private_phf_reexport_for_macro as phf;
 
                 static PHF: phf::Map<&'static str, usize> = phf::phf_map! {
@@ -111,6 +113,9 @@ pub fn documented_fields(input: TokenStream) -> TokenStream {
 #[proc_macro_derive(DocumentedVariants)]
 pub fn documented_variants(input: TokenStream) -> TokenStream {
     let input = parse_macro_input!(input as DeriveInput);
+
+    let ident = &input.ident;
+    let (impl_generics, ty_generics, where_clause) = input.generics.split_for_impl();
 
     let variants_doc_comments = {
         let Data::Enum(DataEnum { variants, .. }) = input.data else {
@@ -132,8 +137,6 @@ pub fn documented_variants(input: TokenStream) -> TokenStream {
             Err(e) => return e.into_compile_error().into(),
         }
     };
-
-    let ident = input.ident;
 
     let match_arms: Vec<_> = variants_doc_comments
         .into_iter()
@@ -159,7 +162,7 @@ pub fn documented_variants(input: TokenStream) -> TokenStream {
     // improvement suggestions are more than welcomed
     quote! {
         #[automatically_derived]
-        impl documented::DocumentedVariants for #ident {
+        impl #impl_generics documented::DocumentedVariants for #ident #ty_generics #where_clause {
             fn get_variant_docs(&self) -> Result<&'static str, documented::Error> {
                 match self {
                     #(#match_arms)*

--- a/documented-test/src/documented.rs
+++ b/documented-test/src/documented.rs
@@ -59,6 +59,54 @@ Note the lack of spaces in front of this line.
 Attribute-style documentation is supported too.";
         assert_eq!(BornIn69::DOCS, doc_str);
     }
+
+    #[test]
+    fn generic_type_works() {
+        /// Wow
+        #[allow(dead_code)]
+        #[derive(Documented)]
+        struct Doge<T> {
+            much: T,
+        }
+
+        assert_eq!(Doge::<u8>::DOCS, "Wow");
+    }
+
+    #[test]
+    fn generic_type_with_bounds_works() {
+        /// Wow
+        #[allow(dead_code)]
+        #[derive(Documented)]
+        struct Doge<T: Copy> {
+            much: T,
+        }
+
+        assert_eq!(Doge::<u8>::DOCS, "Wow");
+    }
+
+    #[test]
+    fn const_generic_type_works() {
+        /// Wow
+        #[allow(dead_code)]
+        #[derive(Documented)]
+        struct Doge<const LEN: usize> {
+            much: [u8; LEN],
+        }
+
+        assert_eq!(Doge::<69>::DOCS, "Wow");
+    }
+
+    #[test]
+    fn lifetimed_type_works() {
+        /// Wow
+        #[allow(dead_code)]
+        #[derive(Documented)]
+        struct Doge<'a> {
+            much: &'a str,
+        }
+
+        assert_eq!(Doge::DOCS, "Wow");
+    }
 }
 
 mod test_qualified {

--- a/documented-test/src/documented_fields.rs
+++ b/documented-test/src/documented_fields.rs
@@ -87,3 +87,51 @@ fn unnamed_fields() {
     assert_eq!(Foo::FIELD_DOCS[1], Some("1"));
     assert_eq!(Foo::FIELD_DOCS[2], None);
 }
+
+#[test]
+fn generic_type_works() {
+    #[derive(DocumentedFields)]
+    #[allow(dead_code)]
+    struct Foo<T> {
+        /// foo
+        foo: T,
+    }
+
+    assert_eq!(Foo::<u8>::get_field_docs("foo"), Ok("foo"));
+}
+
+#[test]
+fn generic_type_with_bounds_works() {
+    #[derive(DocumentedFields)]
+    #[allow(dead_code)]
+    struct Foo<T: Copy> {
+        /// foo
+        foo: T,
+    }
+
+    assert_eq!(Foo::<u8>::get_field_docs("foo"), Ok("foo"));
+}
+
+#[test]
+fn const_generic_type_works() {
+    #[derive(DocumentedFields)]
+    #[allow(dead_code)]
+    struct Foo<const LEN: usize> {
+        /// foo
+        foo: [u8; LEN],
+    }
+
+    assert_eq!(Foo::<69>::get_field_docs("foo"), Ok("foo"));
+}
+
+#[test]
+fn lifetimed_type_works() {
+    #[derive(DocumentedFields)]
+    #[allow(dead_code)]
+    struct Foo<'a> {
+        /// foo
+        foo: &'a u8,
+    }
+
+    assert_eq!(Foo::get_field_docs("foo"), Ok("foo"));
+}

--- a/documented-test/src/documented_variants.rs
+++ b/documented-test/src/documented_variants.rs
@@ -51,3 +51,63 @@ fn works_on_adt_enums() {
         Ok("An empty struct variant.")
     );
 }
+
+#[test]
+fn works_on_generic_enums() {
+    #[allow(dead_code)]
+    #[derive(DocumentedVariants)]
+    enum Foo<T, U> {
+        /// 600
+        Rufus(T),
+        /// 599
+        Dufus(T, U),
+    }
+
+    assert_eq!(Foo::<u8, u8>::Rufus(69).get_variant_docs(), Ok("600"));
+    assert_eq!(Foo::Dufus(69, 420).get_variant_docs(), Ok("599"));
+}
+
+#[test]
+fn works_on_generic_enums_with_bounds() {
+    #[allow(dead_code)]
+    #[derive(DocumentedVariants)]
+    enum Foo<T: Copy, U: std::fmt::Debug> {
+        /// 600
+        Rufus(T),
+        /// 599
+        Dufus(T, U),
+    }
+
+    assert_eq!(Foo::<u8, u8>::Rufus(69).get_variant_docs(), Ok("600"));
+    assert_eq!(Foo::Dufus(69, 420).get_variant_docs(), Ok("599"));
+}
+
+#[test]
+fn works_on_const_generic_enums() {
+    #[allow(dead_code)]
+    #[derive(DocumentedVariants)]
+    enum Foo<const LEN: usize> {
+        /// 600
+        Rufus([u8; LEN]),
+        /// 599
+        Dufus([i8; LEN]),
+    }
+
+    assert_eq!(Foo::Rufus([42; 69]).get_variant_docs(), Ok("600"));
+    assert_eq!(Foo::Dufus([42; 69]).get_variant_docs(), Ok("599"));
+}
+
+#[test]
+fn works_on_lifetimed_enums() {
+    #[allow(dead_code)]
+    #[derive(DocumentedVariants)]
+    enum Foo<'a, T> {
+        /// 600
+        Rufus(&'a T),
+        /// 599
+        Dufus(T, &'a T),
+    }
+
+    assert_eq!(Foo::Rufus(&69).get_variant_docs(), Ok("600"));
+    assert_eq!(Foo::Dufus(69, &420).get_variant_docs(), Ok("599"));
+}

